samza-api/src/main/java/org/apache/samza/operators/windows/BaseWindowFunction.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.samza.operators.windows;
+
+import org.apache.samza.annotation.InterfaceStability;
+import org.apache.samza.operators.data.MessageEnvelope;
+import org.apache.samza.operators.triggers.Trigger;
+
+import java.util.function.BiFunction;
+import java.util.function.Function;
+
+/**
+ *  Implementation of a {@link WindowFunction}. This specifies default triggers for the {@link WindowFunction}, emission
+ *  of early or late results and whether to accumulate or discard previous results.
+ */
+
+@InterfaceStability.Unstable
+public class BaseWindowFunction<M extends MessageEnvelope, K, WV> implements WindowFunction<M, K, WindowKey<K>, WV, WindowOutput<WindowKey<K>, WV>> {
+
+  public enum AccumulationMode { ACCUMULATING, DISCARDING }
+
+  private final Trigger defaultTrigger;
+
+  private Trigger earlyTrigger;
+
+  private Trigger lateTrigger;
+
+  /*
+   * The function that is applied each time a {@link MessageEnvelope} is added to this window.
+   */
+  private final BiFunction<M, WV, WV> foldFunction;
+
+  /*
+   * The function that extracts the key from a {@link MessageEnvelope}
+   */
+  private final Function<M, K> keyExtractor;
```

Comment:
- `WK` is simply a combination of window metadata + group by key `K`. We can provide all information about this pane that we used to fire in the metadata.

- `Function<M,PK>` : Not sure we need this at the moment in the API and hence did not want to add it up-front. Maybe, the `Impl` can infer this / construct this. 

> One more question to be asked: is it enough to find the pane id just from the input message? Do we ever have a case that we need to also rely on existing window state store to determine the pane id an input message belongs to?

Good point, We probably need some look up on the `WindowStateStore` to construct a `paneId`. Eitherways, would n't it be nicer to hide this in the API, and expose this in the implementation classes?